pkg/processor/metricfrequencyprocessor/metriccache.go

```diff
@@ -0,0 +1,55 @@
+package metricfrequencyprocessor
+
+import (
+	"github.com/patrickmn/go-cache"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"time"
+)
+
+type DataPoint struct {
+	Timestamp pdata.Timestamp
+	Value     float64
+}
+
+type MetricCache struct {
+	internalCaches map[string]*cache.Cache
+}
+
+func (mc *MetricCache) Register(name string, dataPoint pdata.DoubleDataPoint) {
+
+	internalCache, found := mc.internalCaches[name]
+	if !found {
+		newCache := mc.newCache()
+		mc.internalCaches[name] = newCache
+		internalCache = newCache
+	}
+
+	key := dataPoint.Timestamp().String()
+	value := &DataPoint{Timestamp: dataPoint.Timestamp(), Value: dataPoint.Value()}
+	internalCache.Set(key, value, cache.DefaultExpiration)
+}
+
+func (mc *MetricCache) List(metricName string) map[pdata.Timestamp]float64 {
+	out := make(map[pdata.Timestamp]float64)
+	internalCache, found := mc.internalCaches[metricName]
+	if found {
+		for _, item := range internalCache.Items() {
+			dataPoint := item.Object.(*DataPoint)
+			out[dataPoint.Timestamp] = dataPoint.Value
+		}
+	}
+
+	return out
+}
+
+func (mc *MetricCache) Cleanup() {
+	for key, internalCache := range mc.internalCaches {
+		if internalCache.ItemCount() == 0 {
+			delete(mc.internalCaches, key)
+		}
+	}
+}
+
+func (mc *MetricCache) newCache() *cache.Cache {
+	return cache.New(time.Hour*1, time.Minute*10)
+}
```

Comment:
This method doesn't really use `MetricCache` so can be a free `func` not attached to the type

Reply:
changed